ultra/ultra/baselines/common/baseline_state_preprocessor.py

```diff
@@ -0,0 +1,231 @@
+# MIT License
+#
+# Copyright (C) 2021. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import collections.abc
+import numpy as np
+import torch
+
+from ultra.baselines.common.state_preprocessor import StatePreprocessor
+from ultra.utils.common import rotate2d_vector
+
+
+class BaselineStatePreprocessor(StatePreprocessor):
+    """The State Preprocessor used by the baseline agents."""
+
+    _NUM_LOOKAHEAD = 100
+    _NORMALIZATION_VALUES = {
+        "speed": 30.0,
+        "distance_from_center": 1.0,
+        "steering": 3.14,  # radians
+        "angle_error": 3.14,  # radians
+        "relative_goal_position": 100.0,
+        # "action": 1.0,  # 2
+        "waypoints_lookahead": 10.0,
+        "road_speed": 30.0,
+    }
+
+    def __init__(
+        self,
+        social_vehicle_config,
+        observation_waypoints_lookahead,
+        action_size,
+    ):
+        self._state_description = self.get_state_description(
+            social_vehicle_config, observation_waypoints_lookahead, action_size
+        )
+
+    @staticmethod
+    def get_state_description(
+        social_vehicle_config, observation_waypoints_lookahead, action_size
+    ):
+        return {
+            "images": {},
```

Comment:
I removed images from all of the RLlib implementations because we don't use them in baseline, can you test train_rllib with this?

Reply:
Okay yes, if we simply uncomment
https://github.com/huawei-noah/SMARTS/blob/17fa18134c658b595b50d79f0d112f50c2064e98/ultra/ultra/baselines/common/baseline_state_preprocessor.py#L89
then we get this error:
```sh
ValueError: ('Observation for a Box/MultiBinary/MultiDiscrete space should be an np.array, not a Python list.', {'images': {}, 'low_dim_states': array([ 0.4453365 ,  0.        , -0.        ,  0.        , -1.2080997 ,
        1.1596714 ,  0.        ,  0.        ,  0.        ,  0.09338313,
        0.        ,  0.18676627,  0.        ,  0.2801494 ,  0.        ,
        0.37353253,  0.        ,  0.46691567,  0.        ,  0.5602988 ,
        0.        ,  0.65368193,  0.        ,  0.74706507,  0.        ,
        0.8404482 ,  0.        ,  0.93383133,  0.        ,  1.0272144 ,
        0.        ,  1.1205976 ,  0.        ,  1.2139807 ,  0.        ,
        1.3073639 ,  0.        ,  1.400747  ,  0.        ,  1.4941301 ,
        0.        ,  1.5875132 ,  0.        ,  1.6808964 ,  0.        ,
        1.7742795 ,  0.463     ], dtype=float32), 'social_vehicles': array([[-3.2000002e-02,  5.2642679e-01, -4.7114328e-01,  5.7698415e-17],
       [ 7.7859737e-02,  1.1286021e+00,  1.2947618e-04, -9.5708752e-01],
       [ 3.1775606e-01,  1.1606046e+00,  4.2924758e-17,  7.0101452e-01],
       [ 6.6240573e-01,  1.0966046e+00, -2.2666438e-16, -1.2339033e+00],
       [ 9.4935733e-01,  1.1606046e+00,  5.5122443e-17,  9.0021777e-01],
       [-1.4803296e+00,  1.0954689e+00,  4.3548187e-04, -5.8546180e-01],
       [-1.4826225e+00,  1.1274673e+00,  4.2566488e-04, -5.7226384e-01],
       [ 0.0000000e+00,  0.0000000e+00,  0.0000000e+00,  0.0000000e+00],
       [ 0.0000000e+00,  0.0000000e+00,  0.0000000e+00,  0.0000000e+00],
       [ 0.0000000e+00,  0.0000000e+00,  0.0000000e+00,  0.0000000e+00]],
      dtype=float32)})
```
If we don't have images to preprocess, then it works if we simply change
https://github.com/huawei-noah/SMARTS/blob/17fa18134c658b595b50d79f0d112f50c2064e98/ultra/ultra/baselines/common/baseline_state_preprocessor.py#L89
to something like `images = np.zeros(1)`. But then if we do this, it may just be better to get rid of the images in the baseline state description, baseline state preprocessor, and baseline adapter, because they aren't used.

I think Alex just wanted to have the options to include images through the Gym wrapper, not necessarily for the baselines. Is this correct @AlexLewandowski? Sorry I could be incorrectly interpreting what you have said.

But anyways, since the baselines don't even use images, I think it would be okay to just remove images from the baseline state description, baseline state preprocessor, and baseline adapter (as long as Alex didn't need them for anything).

Reply:
it's up to you guys if you wanna keep it. I think the images would be more complex as it needs to include dimensions of the image and also consider the number of images to stack. I think it's best to remove images for now and let Alex experiment with them himself instead of including them in baseline

Reply:
Okay great, sounds good.